bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestRackawareEnsemblePlacementPolicy.java

```diff
@@ -1417,6 +1417,176 @@ public boolean validate() {
+            for (int j = 0; j < numOfBookiesPerRack; j++) {
+                int index = i * numOfBookiesPerRack + j;
+                bookieAddress = new BookieSocketAddress("128.0.0." + index, 3181).toBookieId();
+                StaticDNSResolver.addNodeToRack("128.0.0." + index, rackLocationNames[i]);
+                bookieSocketAddresses.add(bookieAddress);
+                bookieRackMap.put(bookieAddress, rackLocationNames[i]);
+            }
+        }
+
+        repp.onClusterChanged(new HashSet<BookieId>(bookieSocketAddresses), new HashSet<BookieId>());
+
+        int writeQuorum = 3;
+        int ackQuorum = 3;
+
+        //test three knows bookie
+        List<BookieId> knowsEnsemble = new ArrayList<>();
+        knowsEnsemble.add(BookieId.parse("128.0.0.0:3181"));
+        knowsEnsemble.add(BookieId.parse("128.0.0.1:3181"));
+        knowsEnsemble.add(BookieId.parse("128.0.0.2:3181"));
+
+        PlacementPolicyAdherence placementPolicyAdherence = repp.isEnsembleAdheringToPlacementPolicy(
+                knowsEnsemble, writeQuorum, ackQuorum);
+        assertEquals(placementPolicyAdherence, PlacementPolicyAdherence.FAIL);
+
+        Map<Integer, BookieId> targetBookie =
+                repp.replaceNotAdheringPlacementPolicyBookie(knowsEnsemble, ackQuorum, writeQuorum,
+                        Collections.emptyMap());
+        //should replace two bookie
+        assertEquals(targetBookie.size(), 2);
+
+        for (Map.Entry<Integer, BookieId> entry : targetBookie.entrySet()) {
+            knowsEnsemble.set(entry.getKey(), entry.getValue());
+        }
+
+        placementPolicyAdherence = repp.isEnsembleAdheringToPlacementPolicy(
+                knowsEnsemble, writeQuorum, ackQuorum);
+        assertEquals(placementPolicyAdherence, PlacementPolicyAdherence.MEETS_STRICT);
+
+        //test three unknowns bookie
+        List<BookieId> unknownEnsembles = new ArrayList<>();
+        unknownEnsembles.add(BookieId.parse("128.0.0.100:3181"));
+        unknownEnsembles.add(BookieId.parse("128.0.0.101:3181"));
+        unknownEnsembles.add(BookieId.parse("128.0.0.102:3181"));
+
+        placementPolicyAdherence = repp.isEnsembleAdheringToPlacementPolicy(
+                unknownEnsembles, writeQuorum, ackQuorum);
+        assertEquals(placementPolicyAdherence, PlacementPolicyAdherence.FAIL);
+
+        //should replace three bookie
+        targetBookie = repp.replaceNotAdheringPlacementPolicyBookie(unknownEnsembles, ackQuorum, writeQuorum,
+                Collections.emptyMap());
+        assertEquals(targetBookie.size(), 3);
+
+        for (Map.Entry<Integer, BookieId> entry : targetBookie.entrySet()) {
+            unknownEnsembles.set(entry.getKey(), entry.getValue());
+        }
+
+        placementPolicyAdherence = repp.isEnsembleAdheringToPlacementPolicy(
+                unknownEnsembles, writeQuorum, ackQuorum);
+        assertEquals(placementPolicyAdherence, PlacementPolicyAdherence.MEETS_STRICT);
```

Comment:
If all ensembles are missing, doesn't that means the entry can not recover anymore?

Reply:
What is `If all ensembles are missing` meaning?